### PR TITLE
generate and install pkg-config file when invoking the install rule

### PR DESCRIPTION
**Pull request by arvidn**

### CODE REVIEW

File: Jamfile [hunk 1]
```diff
@@ -785,6 +785,57 @@ lib torrent
...
+
+	# this is the libtorrent library target
+	local t = [ $(l).find torrent : . ] ;
+
+	# these are the properties we're using to build it with
+	local props = [ $(t).generate [ property-set.create $(properties) ] ] ;
+	props = $(props[1]) ;
+
+	local FULL_VERSION = 1.2.12 ;
+	local p = [ package.paths libtorrent : $(properties) ] ;
```
Comment (emmercm) [inline Jamfile:806]: I had a problem here with `b2 install`, here's the full build output:

```shell
$ mktemp -d

$ cd /tmp/tmp.ntTjfwZK1E

$ git clone https://github.com/arvidn/libtorrent.git
Cloning into 'libtorrent'...

$ cd libtorrent

$ git checkout v1.2.12
Note: switching to 'v1.2.12'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at e3f2b016d update revision
$ git clean --force

$ git submodule update '--depth=1' --init --recursive
Submodule 'simulation/libsimulator' (https://github.com/arvidn/libsimulator.git) registered for path 'simulation/libsimulator'
Cloning into '/tmp/tmp.ntTjfwZK1E/libtorrent/simulation/libsimulator'...
From https://github.com/arvidn/libsimulator
 * branch            6ecbc7edfca9f95ada50d58c54d3de92bfe3890b -> FETCH_HEAD
Submodule path 'simulation/libsimulator': checked out '6ecbc7edfca9f95ada50d58c54d3de92bfe3890b'

$ git apply /5874.patch

$ rm /5026.patch

$ PREFIX=/usr

$ BUILD_CONFIG='release cxxstd=14 crypto=openssl warnings=off address-model=32 -j8'

$ BOOST_ROOT= b2 release 'cxxstd=14' 'crypto=openssl' 'warnings=off' 'address-model=32' -j8 'link=shared' install '--prefix=/usr'
warning: mismatched versions of Boost.Build engine and core
warning: Boost.Build engine (b2) is 4.0.1
warning: Boost.Build core (at /usr/share/boost-build/src) is 2018.02-git
CXXFLAGS =
LDFLAGS =
OS = LINUX
warning: No toolsets are configured.
warning: Configuring default toolset "gcc".
warning: If the default is wrong, your build may not work correctly.
warning: Use the "toolset=xxxxx" option to override our guess.
warning: For more configuration options, please consult
warning: http://boost.org/boost-build2/doc/html/bbv2/advanced/configuration.html
Jamfile:807: in Jamfile</tmp/tmp.ntTjfwZK1E/libtorrent>.generate-pkg-config
ERROR: rule "package.paths" unknown in module "Jamfile</tmp/tmp.ntTjfwZK1E/libtorrent>".
/usr/share/boost-build/src/kernel/modules.jam:107: in modules.call-in
/usr/share/boost-build/src/util/indirect.jam:105: in indirect.call
/usr/share/boost-build/src/build/property.jam:144: in property.evaluate-conditionals-in-context
/usr/share/boost-build/src/build/targets.jam:1087: in evaluate-requirements
/usr/share/boost-build/src/build/targets.jam:1121: in common-properties2
/usr/share/boost-build/src/build/targets.jam:1017: in targets.common-properties
/usr/share/boost-build/src/build/targets.jam:1313: in class@basic-target.generate
/usr/share/boost-build/src/build/targets.jam:812: in generate-really
/usr/share/boost-build/src/build/targets.jam:784: in class@main-target.generate
/usr/share/boost-build/src/build/targets.jam:933: in targets.generate-from-reference
/usr/share/boost-build/src/build/targets.jam:1276: in generate-dependencies
/usr/share/boost-build/src/build/targets.jam:1333: in class@basic-target.generate
/usr/share/boost-build/src/build/targets.jam:812: in generate-really
/usr/share/boost-build/src/build/targets.jam:784: in class@main-target.generate
/usr/share/boost-build/src/build-system.jam:797: in load
/usr/share/boost-build/src/kernel/modules.jam:295: in import
/usr/share/boost-build/src/kernel/bootstrap.jam:139: in boost-build
/usr/share/boost-build/boost-build.jam:8: in module scope
```

Unsure if this PR is more intended for v2.0 going forward or if it's possible to use with v1.2 - could be a bad assumption from me.

File: tools/set_version.py [hunk 2]
```diff
@@ -49,6 +49,8 @@ def substitute_file(name):
             line = '    version="%d.%d.%d",\n' % (version[0], version[1], version[2])
         elif '"-LT' in line and name.endswith('settings_pack.cpp'):
             line = re.sub('"-LT[0-9A-Za-z]{4}-"', '"-LT%c%c%c%c-"' % v(version), line)
+        elif 'local FULL_VERSION = ' in line and name == 'Jamfile':
+            line = '\tlocal FULL_VERSION = %d.%d.%d ;\n' % (version[0], version[1], version[2])
```
Comment (emmercm) [inline tools/set_version.py:53]: Actually, I probably answered my own question, this didn't work with v2.0.1 for me:

```
error: patch failed: tools/set_version.py:49
error: tools/set_version.py: patch does not apply
```

**Comment by arvidn** [comment]
> ERROR: rule "package.paths" unknown in module "Jamfile</tmp/tmp.ntTjfwZK1E/libtorrent>".

This suggests that the version of boost-build (`b2`) you're using is too old. I should probably test the install rule on CI as well, to ensure I'm not depending on a too new version. I'll take a look at when this was introduced.

File: Jamfile (hunk 1 above)
Comment (arvidn) [reply Jamfile:806]: I suspect this is a hint of the root cause:

```
warning: mismatched versions of Boost.Build engine and core
warning: Boost.Build engine (b2) is 4.0.1
warning: Boost.Build core (at /usr/share/boost-build/src) is 2018.02-git
```

File: tools/set_version.py (hunk 2 above)
Comment (arvidn) [reply tools/set_version.py:53]: it's likely some additional patches on top of `v1.2.12` (that are in `RC_1_2`) prevent this part of the patch from applying. This is not important though

File: Jamfile (hunk 1 above)
Comment (arvidn) [reply Jamfile:806]: this message suggests that the `boost-build.jam` files (that implement the build system) and the engine `b2` are from different versions. This sometimes causes errors. Are both `b2` and `/usr/share/boost-build` from the same package?

**Comment by arvidn** [comment]
got the error on ubuntu 18: https://github.com/arvidn/libtorrent/pull/5874/checks?check_run_id=1724236138#step:7:893

**Comment by arvidn** [comment]
@emmercm I've updated this branch now. it works on older ubuntu as well

**Comment by zywo** [comment]
```pkgconfig``` is installed in ```${PREFIX}/share``` not in ```${PREFIX}/lib```

**Comment by arvidn** [comment]
> pkgconfig is installed in ${PREFIX}/share not in ${PREFIX}/lib

That's right

**Comment by arvidn** [comment]
can anyone confirm that this seems correct?
@emmercm does this enable qbt to pick up the library?

**Comment by emmercm** [comment]
@arvidn yep testing this tonight, thanks so much for your work on this!

File: Jamfile (hunk 1 above)
Comment (emmercm) [reply Jamfile]: Looks like no Alpine package has a `boost-build` file, but `b2` is v1.72.0-r6 according to https://pkgs.alpinelinux.org/package/v3.12/main/x86_64/boost.

Comment (emmercm) [reply Jamfile]: You're right, `boost-build` is v1.69.0-r1: https://pkgs.alpinelinux.org/package/v3.12/community/x86_64/boost-build.

**Comment by emmercm** [comment]
Sorry, unclear message from me there, I intended to say I'd test this on my end tonight - unfortunately I'm still getting `ERROR: rule "package.paths" unknown in module "Jamfile</tmp/tmp.wDzOIMabGH/libtorrent>".`:

```shell
$ mktemp -d

$ cd /tmp/tmp.wDzOIMabGH

$ git clone https://github.com/arvidn/libtorrent.git
Cloning into 'libtorrent'...

$ cd libtorrent

$ git checkout v1.2.12
Note: switching to 'v1.2.12'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at e3f2b016d update revision

$ git clean --force

$ git submodule update '--depth=1' --init --recursive
Submodule 'simulation/libsimulator' (https://github.com/arvidn/libsimulator.git) registered for path 'simulation/libsimulator'
Cloning into '/tmp/tmp.wDzOIMabGH/libtorrent/simulation/libsimulator'...
From https://github.com/arvidn/libsimulator
 * branch            6ecbc7edfca9f95ada50d58c54d3de92bfe3890b -> FETCH_HEAD
Submodule path 'simulation/libsimulator': checked out '6ecbc7edfca9f95ada50d58c54d3de92bfe3890b'

$ git apply /5026.patch

$ rm /5026.patch

$ git apply /5874.patch

$ rm /5874.patch

$ BUILD_CONFIG='release cxxstd=14 crypto=openssl warnings=off address-model=32 -j8'

$ BOOST_ROOT= b2 release 'cxxstd=14' 'crypto=openssl' 'warnings=off' 'address-model=32' -j8 'link=shared' install '--prefix=/usr'
warning: mismatched versions of Boost.Build engine and core
warning: Boost.Build engine (b2) is 4.0.1
warning: Boost.Build core (at /usr/share/boost-build/src) is 2018.02-git
CXXFLAGS =
LDFLAGS =
OS = LINUX
warning: No toolsets are configured.
warning: Configuring default toolset "gcc".
warning: If the default is wrong, your build may not work correctly.
warning: Use the "toolset=xxxxx" option to override our guess.
warning: For more configuration options, please consult
warning: http://boost.org/boost-build2/doc/html/bbv2/advanced/configuration.html
Jamfile:833: in Jamfile</tmp/tmp.wDzOIMabGH/libtorrent>.generate-pkg-config
ERROR: rule "package.paths" unknown in module "Jamfile</tmp/tmp.wDzOIMabGH/libtorrent>".
/usr/share/boost-build/src/kernel/modules.jam:107: in modules.call-in
/usr/share/boost-build/src/util/indirect.jam:105: in indirect.call
/usr/share/boost-build/src/build/property.jam:144: in property.evaluate-conditionals-in-context
/usr/share/boost-build/src/build/targets.jam:1087: in evaluate-requirements
/usr/share/boost-build/src/build/targets.jam:1121: in common-properties2
/usr/share/boost-build/src/build/targets.jam:1017: in targets.common-properties
/usr/share/boost-build/src/build/targets.jam:1313: in class@basic-target.generate
/usr/share/boost-build/src/build/targets.jam:812: in generate-really
/usr/share/boost-build/src/build/targets.jam:784: in class@main-target.generate
/usr/share/boost-build/src/build/targets.jam:933: in targets.generate-from-reference
/usr/share/boost-build/src/build/targets.jam:1276: in generate-dependencies
/usr/share/boost-build/src/build/targets.jam:1333: in class@basic-target.generate
/usr/share/boost-build/src/build/targets.jam:812: in generate-really
/usr/share/boost-build/src/build/targets.jam:784: in class@main-target.generate
/usr/share/boost-build/src/build-system.jam:797: in load
/usr/share/boost-build/src/kernel/modules.jam:295: in import
/usr/share/boost-build/src/kernel/bootstrap.jam:139: in boost-build
/usr/share/boost-build/boost-build.jam:8: in module scope
```

File: Jamfile (hunk 1 above)
Comment (emmercm) [reply Jamfile]: I'm actually using Alpine v3.13 now, but it looks like they haven't updated the Boost packages from v3.12, so the above are still correct.

**Comment by arvidn** [comment]
It seems boost-1.69 and boost-1.70 used the same version number for boost-build, so I can distinguish this with that precision. Please try with this patch:

```
diff --git a/Jamfile b/Jamfile
index 1735ef241..8eeff1145 100644
--- a/Jamfile
+++ b/Jamfile
@@ -811,7 +811,7 @@ rule generate-pkg-config ( properties * )
        # however, boost build's versioning scheme changed in boost-1.71 to version
        # 4.0
        local boost-build-version = [ SPLIT_BY_CHARACTERS [ version.boost-build ] : "-" ] ;
-       if [ version.version-less [ SPLIT_BY_CHARACTERS $(boost-build-version[1]) : "." ] : 2018 01 ]
+       if [ version.version-less [ SPLIT_BY_CHARACTERS $(boost-build-version[1]) : "." ] : 2018 03 ]
        {
                import option ;
                import property ;

```

**Comment by arvidn** [comment]
also here: https://github.com/arvidn/libtorrent/pull/5894

**Pull request merged**